tools/find_device.c

```diff
@@ -0,0 +1,146 @@
+/*
+ * Copyright 2020 Codethink Ltd.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <dirent.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <libudev.h>
+#include <locale.h>
+#include <unistd.h>
+#include <stdbool.h>
+
+#include "find_device.h"
+
+static int match(char *sysname)
+{
+    struct udev *udev;
+    struct udev_enumerate *enumerate;
+    struct udev_list_entry *device;
+    struct udev_device *dev;
+    const char *path;
+    bool ret = false;
+
+    /* Create the udev object */
+    udev = udev_new();
+    if (!udev)
+    {
+        printf("Can't create udev\n");
+        exit(1);
+    }
+
+    /* Create a list of the devices with the name "ttyUSB*". */
+    enumerate = udev_enumerate_new(udev);
+    udev_enumerate_add_match_sysname(enumerate, sysname);
+    udev_enumerate_scan_devices(enumerate);
+
+    /* udev_enumerate_get_list_entry returns a list of entry,
+    but we only need the first entry, because we only cares
+    about their parent which are the same for all. */
+    device = udev_enumerate_get_list_entry(enumerate);
+
+    /* Get the filename of the /sys entry for the device
+     and create a udev_device object (dev) representing it */
+    path = udev_list_entry_get_name(device);
+    dev = udev_device_new_from_syspath(udev, path);
+
+    /* In order to get information about the
+     USB device, get the parent device with the
+     subsystem/devtype pair of "usb"/"usb_device". This will
+     be several levels up the tree, but the function will find
+     it.*/
+    dev = udev_device_get_parent_with_subsystem_devtype(
+        dev,
+        "usb",
+        "usb_device");
+    if (!dev)
+    {
+        printf("Unable to find parent usb device.");
+        exit(1);
+    }
+
+    /* From here, we can call get_sysattr_value() for each file
+     in the device's /sys entry. The strings passed into these
+     functions (idProduct, idVendor, serial, etc.) correspond
+     directly to the files in the directory which represents
+     the USB device. Note that USB strings are Unicode, UCS2
+     encoded, but the strings returned from
+     udev_device_get_sysattr_value() are UTF-8 encoded. */
+    if (!strncmp(udev_device_get_sysattr_value(dev, "manufacturer"),
+                 "Codethink", 9) &&
+        !strncmp(udev_device_get_sysattr_value(dev, "product"),
+                 "Medical Plethysmograph Device", 29))
```

Comment:
This is cool.

I think we should put the strings like "Medical Plethysmograph Device" into a header like:

```c
#define BL_STR_PRODUCT "Medical Plethysmograph Device"
```

And use the `BL_STR_PRODUCT` here and in `usb.c` to keep them in sync.

Reply:
+1